eva_submission/vep_utils.py

```diff
@@ -0,0 +1,223 @@
+    genome_ftp = get_ftp_connection(ensembl_genome_ftp_url)
+    for subdir in ensembl_genome_dirs:
+        genome_releases = get_releases(genome_ftp, subdir, current_release_only)
+        release = search_releases(genome_ftp, genome_releases, species_name, assembly_name)
+        if release:
+            return release, 'genomes'
+
+    logger.info(f'No VEP cache found anywhere on Ensembl FTP for {species_name} and {assembly_name}')
+    return None, None
+
+
+@retry(tries=4, delay=2, backoff=1.2, jitter=(1, 3))
+def get_species_and_assembly(taxonomy_id):
+    """
+    Returns Ensembl species name, assembly name and accession of the current supported assembly,
+    or None if any aren't found.
+    """
+    url = f'https://rest.ensembl.org/info/genomes/taxonomy/{taxonomy_id}?content-type=application/json'
+    response = requests.get(url)
+    # This endpoint returns 500 even if taxon id is valid but just not in Ensembl, to minimise user frustration
+    # we'll assume the species is just not currently supported.
+    if not response.ok:
+        logger.warn(f'Got {response.status_code} when trying to get species and assembly from Ensembl.')
+        return None, None, None
+    json_response = response.json()[0]
+    return json_response['name'], json_response['assembly_name'], json_response['assembly_accession']
+
+
+@retry(tries=4, delay=2, backoff=1.2, jitter=(1, 3), logger=logger)
+def get_ftp_connection(url):
+    ftp = ftplib.FTP(url)
+    ftp.login()
+    return ftp
+
+
+@retry(tries=8, delay=2, backoff=1.2, jitter=(1, 3), logger=logger)
+def get_releases(ftp, subdir, current_only):
+    """
+    Get all release version numbers and paths to the releases starting from a given subdirectory.
+    If current_only is True it will only return the most recent release, which can be helpful if the FTP is unreliable.
+    """
+    all_releases = {}
+    for file in ftp.nlst(subdir):
+        if "release-" in file:
+            release_number = file[file.index("-") + 1:]
+            all_releases.update({int(release_number): file})
+    if current_only:
+        current = max(all_releases)
+        logger.info(f'Only getting the most recent release: {current}')
+        return {current: all_releases[current]}
+    return all_releases
+
+
+def search_releases(ftp, all_releases, species, assembly):
+    for release in sorted(all_releases, reverse=True):
+        logger.info(f'Looking for vep_cache_version in release : {all_releases[release]}')
+        all_species_files = get_all_species_files(ftp, all_releases.get(release))
+        for f in all_species_files:
+            if species in f and assembly in f:
+                logger.info(f'Found vep_cache_version for {species} and {assembly}: file {f}, release {release}')
+                # TODO assume if we get here we need to download the cache... is this correct?
+                #  e.g. what if we've downloaded the cache for another study but VEP annotation step failed...
```

Comment:
This is a fair concern which we could address in `get_vep_and_vep_cache_version` by only downloading a VEP cache if it does not exist already. This said downloading it again and overwriting it does not sound too onerous so it should be fine as it is now.